moved objectManager to /js folder; implemented ImageMapType

TNT0305 opened the pull request:
I'm not 100% on if this implementation is imageMapType is good (kinda specific url format), but its a start.  The important part is adding the eval() fallback in the tryParseJson.  This allows us to include functions in the json string we use to initialize the imageMapType.  I also bumped things to .net 5 (even though .net 6 is is out, now) so we could employ js isloation.  I started trying to implement loading of google maps js in the component library, but I ran into CORS issues (what else is new).

+employed js isolation so users don't not have to reference the .js file, themselves.
+Added ImageMapType and supporting js for show/hide layers
Server-side demo changed to .net 5 (for js isolation)
_Host.cshtml removed reference to objectManager.js
"Events" page of server side demo works (Because it uses the GoogleMap root component rather than the Map component)!

valentasm1 commented on GoogleMapsComponents/GoogleMapsComponents.csproj:
```diff
@@ -8,7 +8,7 @@
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>net5.0</TargetFramework>
```
It should stay netstandart2.0. At most you could increase to 2.1

TNT0305 replied:
I was going for the js isolation.  I thought that wasn't included in 2.1.
Regardless.  The js isolation was just a "nice-to-have" that I thought I'd include in this PR.  The eval in the tryParseJson along with the ImageMapType are the functional changes.

valentasm1 commented:
Please create demo page on any of demo solution. Aslo do you have google js example?
I strongly disagree with moduleImports rest i have to check it.

TNT0305 commented:
I checked that the "Events" page on the server side demo worked (after changing it, also, to .net 5, to work with my change to .net 5 in the component library).  I didn't want to go _too_ far down making a full-fledged example project until I got some feedback from you on if these changes would be something you'd want in there.
I _really_ like the js isolation for this as a consumer of the library wouldn't really care what is involved in getting maps going...in general, they just want to include the component and go; however, I certainly undesrstand wanting to keep the component library as .net standard so framework users can still consume it.

valentasm1 commented:
It is not if working now. netstandart give most widest support for any type of platform. What if maps would be run in arm device. It is not debate about this. 
js isolation dont bring any big feature. I agree it is nice to have feature, but most blazor libraries work that you need to provide urls
I love PR. Please create demo page and i will check functionality.

valentasm1 commented:
Is it that ImageMapType  only works with only .net5?

TNT0305 commented:
The ImageMapType should work in netstandard2.0 mode.  I only went to .net 5 for the js isolation.  
Maybe I should do PR with only functional changes and then different PR for js isolation.

TNT0305 commented on GoogleMapsComponents/Maps/ImageMapType.cs:
```diff
@@ -1,11 +1,87 @@
+        public Guid Guid => _jsObjectRef.Guid;
+        public string Name { get; private set; }
+        GoogleMap map = null;
+        public async static Task<ImageMapType> CreateAsync(IJSRuntime jsRuntime, string baseUrl, int minZoom, int maxZoom, string name, float opacity)
+        {
+            string initOpts = @"{
+                'getTileUrl': (coord, zoom) => {
+                            return '" + baseUrl + @"' + zoom + '/' + coord.x + '/' + coord.y;
+                        },
+                'tileSize': new google.maps.Size(256, 256),
+                'maxZoom': " + maxZoom.ToString() + @",
+                'minZoom': " + minZoom.ToString() + @",
+                'opacity': " + opacity.ToString() + @",
+                'name': '" + name + @"'
+            }";
+
+            var jsObjectRef = await JsObjectRef.CreateAsync(jsRuntime, "google.maps.ImageMapType", initOpts);
+            var to = new ImageMapType(jsObjectRef) {
+                Name = name
+            };
+            return to;
+        }
```
this form of getTileUrl is a typical scenario; however, if we could figure a way to set this up with a ImageMapOptions so that a user could really customize their Url, that would be good, I think.

rungwiroon commented:
Should we create branch and set to new version (like 2.0-alpha) and target to .NET 6 (LTS) to support a lot of new blazor features?

TNT0305 commented:
For now, I just checked in changes to return component library to netstandard2.0 (and removed js isolation code)
In the serverSample project, I made a sample of using ImageMapType using OSM base tiles.  I included an opacity slider.  I changed up the ImageMapType to do replacements (like, for OSM, I now pass in: "https://a.tile.openstreetmap.org/{z}/{x}/{y}.png" and ImageMapType is able to make the correct function for google to grab all the right tiles)

valentasm1 commented:
> Should we create branch for new version (like 2.0-alpha) and target to .NET 6 (LTS) to support a lot of new blazor features?

Like what features? Js isolation isnt big, cool feature in this case

TNT0305 commented:
FYI, it is probably a bad idea to release on github something that uses some "other" base map tiles on top of google maps (google license issue, I think).  I wanted to ensure _you_ were able to see overlay tiles, though.

valentasm1 commented:
> FYI, it is probably a bad idea to release on github something that uses some "other" base map tiles on top of google maps (google license issue, I think). I wanted to ensure _you_ were able to see overlay tiles, though.

Maybe yes maybe no. I think google earn from maps from api request limits and functionality. Also i know you should cache google requests for very long (or at all) so i dont think this is issue, but i could be wrong.

valentasm1 commented:
Also if you know any cool features from .net 6 which would be very usefull here i dont mind migrating to it. Js isolation (especially in this case) doesnt sound like one.

TNT0305 commented:
Yes. JS isolation is just "nice", but not "necessary".  Other than that, net6 is LTS.  If/when I encounter "must have" features for this library, I will let you know.

TNT0305 commented:
From google license agreement:

> (e) No Use With Non-Google Maps. To avoid quality issues and/or brand confusion, Customer will not use the Google Maps Core Services with or near a non-Google Map in a Customer Application. For example, Customer will not (i) display or use Places content on a non-Google map, (ii) display Street View imagery and non-Google maps on the same screen, or (iii) link a Google Map to non-Google Maps content or a non-Google map.

So, we should probably _not_ use OSM as sample overlay -- even on github sample (just to be safe)

rungwiroon commented:
> 
> 
> > Should we create branch for new version (like 2.0-alpha) and target to .NET 6 (LTS) to support a lot of new blazor features?
> 
> Like what features? Js isolation isnt big, cool feature in this case

- C# 10
- AsyncEnumerable
- More JS interop API
- Performance Improvement (benchmark?)
- Replace Json.NET with System.Text.Json

TNT0305 commented:
I agree.  C#10 and AsyncEnumerable are very useful.  But, those are for just "inside" component library.  Is it better to be more flexible in component library code or to usuable by more people?  (All my stuff with this is new code, but I understand wanting to use it from older cold)

valentasm1 commented:
@rungwiroon ok ok. Got your point. 
I still disagree target only .net6 but agree that target both platforms like @TNT0305  suggest would be good idea.

valentasm1 commented:
Couple examples of popular blazor nuget and check what they target.
So i would say 50/50.
https://www.nuget.org/packages/Radzen.Blazor/
https://www.nuget.org/packages/MatBlazor/
https://www.nuget.org/packages/AntDesign/
https://www.nuget.org/packages/Blazored.Menu/

rungwiroon commented:
@valentasm1 Everyone use .net standard 2.1 at minimum. Should we just change to .net standard 2.1 for now?

TNT0305 commented:
One thing to consider.  Most people using this blazor component will be at least .net 5, right?  In short term, it would only be stopping early adopters of blazor who still use core 3.x.  But, I think those people should update to at least net 5.  (I do not know if we should force them to, though, but netstandard will not advance-- it was only temporary for going between framework and core)

rungwiroon commented:
OK. I will stick with .net standard 2.1 until .net core 3.1 end of support, after that I can migrate to .net 6 completely (maybe at begin of 2023)

valentasm1 commented:
Regarding migrations/upgrade created ticket. If you have question or something to say regarding it lets discuss there.
https://github.com/rungwiroon/BlazorGoogleMaps/issues/161

valentasm1 commented:
I am having some issues since i updated to .net6 so not all old solutions works for me. Will check in couple days and will update here.

valentasm1 commented:
It doesnt work. I created a branch. What i am doing wrong.
Also only now understand using OSM for google maps. I agree that it could make legal issues. Just for demo purhoses you could leave when i will test functionality i will delete branch and rest related.
https://github.com/rungwiroon/BlazorGoogleMaps/tree/imagemap

valentasm1 commented:
Have you tried fixing example?

TNT0305 commented:
Sorry for the extreme delay.  In the server project, the page route was changed to kabob-case and the navMenu still had the camelCase route.  I corrected that and now the image overlay demo in the server app works.

TNT0305 commented:
oops...I realized I was unable to push to the branch. Do I need to do a PR on the branch?

valentasm1 commented:
Just give me commit somewhere and i will merge by hand or somehow :).

TNT0305 commented:
hmm...I had tried to commit on the imagemap branch...I don't have permission, it seems:
![image](https://user-images.githubusercontent.com/5560099/143899184-f53d0073-a462-42ec-97a9-8d7be427f1f6.png)

TNT0305 commented:
Oh, I see what you mean...the code is in NavMenu.razor:

```razor
        <li class="nav-item px-3">
            <NavLink class="nav-link" href="map-image-overlay">
                <span class="oi oi-list-rich" aria-hidden="true"></span> Image Overlay
            </NavLink>
        </li>
```
the href needed to change

TNT0305 commented:
Here's a screenshot of it running with that change (I've highlighted the opacity slider...more to the right is more OSM, and more to the left is more google):
![image](https://user-images.githubusercontent.com/5560099/143901054-65378666-cfa8-43c1-a588-9aecc9976e35.png)

valentasm1 commented:
So you just clone imagemap branch and you could run server side demo?

TNT0305 commented:
That is correct.
![image](https://user-images.githubusercontent.com/5560099/143912521-2de220e5-7e6a-4fc8-8f3d-f405fed98fa7.png)

valentasm1 commented:
It just dont work for me. Giving same exception
I will try to release in coming days and finger crossed that wont break rest stuff.
```
blazor.server.js:19 [2021-11-29T18:47:51.827Z] Error: Microsoft.JSInterop.JSException: Cannot read properties of undefined (reading 'call')
TypeError: Cannot read properties of undefined (reading 'call')
    at kaa (https://maps.googleapis.com/maps/api/js?key=AIzaSyBdkgvniMdyFPAcTlcZivr8f30iU-kn1T0&v=3&libraries=places,visualization,drawing:50:410)
    at new Ii (https://maps.googleapis.com/maps/api/js?key=AIzaSyBdkgvniMdyFPAcTlcZivr8f30iU-kn1T0&v=3&libraries=places,visualization,drawing:177:164)
    at Object.createObject (https://localhost:44356/_content/BlazorGoogleMaps/js/objectManager.js:295:19)
    at https://localhost:44356/_framework/blazor.server.js:8:35886
    at new Promise (<anonymous>)
    at e.beginInvokeJSFromDotNet (https://localhost:44356/_framework/blazor.server.js:8:35854)
    at https://localhost:44356/_framework/blazor.server.js:1:20247
    at Array.forEach (<anonymous>)
    at e.invokeClientMethod (https://localhost:44356/_framework/blazor.server.js:1:20217)
    at e.processIncomingData (https://localhost:44356/_framework/blazor.server.js:1:18028)
   at Microsoft.JSInterop.JSRuntime.InvokeWithDefaultCancellation[T](String identifier, Object[] args)
   at GoogleMapsComponents.Helper.MyInvokeAsync[TRes](IJSRuntime jsRuntime, String identifier, Object[] args) in C:\Dev\My\BlazorGoogleMaps\GoogleMapsComponents\Helper.cs:line 174
   at GoogleMapsComponents.JsObjectRef.CreateAsync(IJSRuntime jsRuntime, Guid guid, String functionName, Object[] args) in C:\Dev\My\BlazorGoogleMaps\GoogleMapsComponents\JsObjectRef.cs:line 124
   at GoogleMapsComponents.Maps.ImageMapType.CreateAsync(IJSRuntime jsRuntime, String baseUrlFormat, Int32 minZoom, Int32 maxZoom, String name, Single opacity) in C:\Dev\My\BlazorGoogleMaps\GoogleMapsComponents\Maps\ImageMapType.cs:line 43
   at ServerSideDemo.Pages.MapImageOverlay.OnAfterInitAsync() in C:\Dev\My\BlazorGoogleMaps\ServerSideDemo\Pages\MapImageOverlay.razor:line 101
   at Microsoft.AspNetCore.Components.ComponentBase.CallStateHasChangedOnAsyncCompletion(Task task)
   at GoogleMapsComponents.GoogleMap.OnAfterRenderAsync(Boolean firstRender) in C:\Dev\My\BlazorGoogleMaps\GoogleMapsComponents\GoogleMap.razor:line 67
   at Microsoft.AspNetCore.Components.RenderTree.Renderer.GetErrorHandledTask(Task taskToHandle)
```

TNT0305 commented:
I'm not sure what the problem is.  I deleted the entire direcotry and re-cloned it.  I then changed vs to imagemap branch.  At this point, vs complainged about the net5 client.  It turns out, there's a "global.json" that is indicating 3.1.415.  While that was there, things didn't build properly.  I removed the net5 project and it worked.  I also tried it with only removing the global.json file and that also worked.  It might be that you're getting "build weirdness" because of that global.json.  Try deleting that file and do a full rebuild

valentasm1 commented:
https://www.nuget.org/packages/BlazorGoogleMaps/1.3.0
Let me know if it works.
